packages/history/src/index.ts

```diff
@@ -117,52 +151,69 @@ export function createHistory(opts: {
     },
     push: (path, state, navigateOpts) => {
       state = assignKey(state)
-      tryNavigation(() => {
-        opts.pushState(path, state)
-        notify()
-      }, navigateOpts)
+      tryNavigation({
+        task: () => {
+          opts.pushState(path, state)
+          notify()
+        },
+        navigateOpts,
+        type: 'PUSH',
+        path,
+        state,
+      })
     },
     replace: (path, state, navigateOpts) => {
       state = assignKey(state)
-      tryNavigation(() => {
-        opts.replaceState(path, state)
-        notify()
-      }, navigateOpts)
+      tryNavigation({
+        task: () => {
+          opts.replaceState(path, state)
+          notify()
+        },
+        navigateOpts,
+        type: 'REPLACE',
+        path,
+        state,
+      })
     },
     go: (index, navigateOpts) => {
-      tryNavigation(() => {
-        opts.go(index)
-        notify()
-      }, navigateOpts)
+      tryNavigation({
+        task: () => {
+          opts.go(index)
+          notify()
+        },
+        navigateOpts,
+        type: 'POP',
+      })
     },
     back: (navigateOpts) => {
-      tryNavigation(() => {
-        opts.back()
-        notify()
-      }, navigateOpts)
+      tryNavigation({
+        task: () => {
+          opts.back(navigateOpts?.ignoreBlocker ?? false)
+          notify()
+        },
+        navigateOpts,
+        type: 'POP',
+      })
     },
     forward: (navigateOpts) => {
-      tryNavigation(() => {
-        opts.forward()
-        notify()
-      }, navigateOpts)
+      tryNavigation({
+        task: () => {
+          opts.forward(navigateOpts?.ignoreBlocker ?? false)
+          notify()
+        },
+        navigateOpts,
+        type: 'POP',
+      })
     },
     createHref: (str) => opts.createHref(str),
     block: (blocker) => {
-      blockers.push(blocker)
-
-      if (blockers.length === 1) {
-        addEventListener(beforeUnloadEvent, beforeUnloadListener, {
-          capture: true,
-        })
-      }
+      if (!opts.win) return () => {}
+      if (!opts.win.blockers) opts.win.blockers = []
+      opts.win.blockers.push(blocker)
```

Comment:
why do we need to store the blockers in the window?

Reply:
We dont :)